core/network/src/discovery.rs

```diff
@@ -181,84 +181,83 @@ where
 			Self::OutEvent,
 		>,
 	> {
-		// Poll Kademlia.
-		match self.kademlia.poll(params) {
-			Async::NotReady => (),
-			Async::Ready(NetworkBehaviourAction::GenerateEvent(ev)) => {
-				match ev {
-					KademliaOut::Discovered { .. } => {}
-					KademliaOut::KBucketAdded { peer_id, .. } => {
-						let ev = DiscoveryOut::Discovered(peer_id);
-						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-					}
-					KademliaOut::FindNodeResult { key, closer_peers } => {
-						trace!(target: "sub-libp2p", "Libp2p => Query for {:?} yielded {:?} results",
-							key, closer_peers.len());
-						if closer_peers.is_empty() {
-							warn!(target: "sub-libp2p", "Libp2p => Random Kademlia query has yielded empty \
-								results");
-						}
-					}
-					KademliaOut::GetValueResult(res) => {
-						let ev = match res {
-							GetValueResult::Found { results } => {
-								let results = results
-										.into_iter()
-										.map(|r| (r.key, r.value))
-										.collect();
-
-								DiscoveryOut::ValueFound(results)
-							}
-							GetValueResult::NotFound { key, .. } => {
-								DiscoveryOut::ValueNotFound(key)
-							}
-						};
-						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-					}
-					KademliaOut::PutValueResult(res) => {
-						let ev = match res {
-							PutValueResult::Ok{ key, .. } =>  {
-								DiscoveryOut::ValuePut(key)
-							}
-							PutValueResult::Err { key, .. } => {
-								DiscoveryOut::ValuePutFailed(key)
-							}
-						};
-						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-					}
-					// We never start any other type of query.
-					KademliaOut::GetProvidersResult { .. } => {}
-				}
+		// Poll the stream that fires when we need to start a random Kademlia query.
+		match self.next_kad_random_query.poll() {
```

Comment:
I would keep the loop here as well. It probably won't change the behaviour in practice (Kademlia events should wake up the task), but I would keep it for correctness and consistency.

Reply:
I'm not sure I understand yet. As far as I could see, 2 of the 3 code paths of the pattern match had a `break` before, and the third resets `next_kad_random_query` using an instant in the future, in which case polling it again right after doesn't seem to make sense. What am I missing?

Reply:
> in which case polling it again right after doesn't seem to make sense

If you don't poll it again right after, it will not wake up the task once it's ready.

(I may be mistaken and it's actually implemented in a way that it does, but that would be an implementation detail, and not something we should rely on)

Reply:
Ok, done. I was indeed assuming that `reset` registers the current task for notification if the new deadline has not elapsed, but I didn't find evidence for that either, so better be safe than sorry.